solr/solrj/src/java/org/apache/solr/common/cloud/DocCollection.java

```diff
@@ -156,14 +181,37 @@ public static String getCollectionPathRoot(String coll) {
 
   /**
    * Update our state with a state of a {@link Replica} Used to create a new Collection State when
-   * only a replica is updated
+   * only PerReplicaStates are updated
    */
   public DocCollection copyWith(PerReplicaStates newPerReplicaStates) {
-    if (this.prsSupplier != null) {
-      log.debug("In-place update of PRS: {}", newPerReplicaStates);
-      this.prsSupplier.prs = newPerReplicaStates;
+    if (newPerReplicaStates != null) {
+      log.debug("Copy with newPerReplicaStates: {}", newPerReplicaStates);
+      return new DocCollection(name, slices, propMap, router, znodeVersion, newPerReplicaStates);
+    } else {
+      return this;
     }
-    return this;
+  }
+
+  private static Map<String, Slice> mergePerReplicaStatesToSlices(
```

Comment:
I'm concerned about the no:of objects we create on every PRS change. Currently, no new Objects are created as the state is fetched in real-time

Reply:
Yes, I agree with the saved object instance and this is just a proposal.

I guess this is the trade-off : if DocCollection is to be immutable then we cannot re-use the `DocCollection` instance.

This is the existing design of classes such as `DocCollection`, `Slice` and `Replica`. Making them mutable could risk stability of code elsewhere that relies on those object's state being consistent and stable after the instance is obtained :)

Reply:
@noblepaul I gave it another stab to do a deeper refactoring that removes `PerReplicaStates` from `DocCollection` completely. As shown in https://github.com/cowpaths/fullstory-solr/pull/84

But again, I haven't run the full test cases yet. The basic tests seem to be passing but im sure there will be some failing (will look at them later if we want to investigate more).

This is mostly just exploratory work and see how we feel about it.

If it is a direction we want to take more serious consideration, i'm all for fixing all the test cases/failures and open a jira ticket in Apache solr 😊 

Thanks again!